### PR TITLE
[PYTHON API] fix direct access to model input shape

### PR DESCRIPTION
**Pull request by AlexeyLebedev1**
### Details:
 - *Input shapes getters and properties return constant reference from cpp, but python casts it to non constant, it leads to python allow write to input shape buffer directly using `__getitem__`, that makes impossible to reshape model after that. Solution - return copy of shape instead reference in python.*
 - *Bind `__eq__` for `Shape`*

### CODE REVIEW

File: src/bindings/python/src/pyopenvino/graph/node_output.hpp [hunk 1]
```diff
@@ -78,6 +78,7 @@ void regclass_graph_Output(py::module m, std::string typestring)
                )");
     output.def("get_shape",
                &ov::Output<VT>::get_shape,
+               py::return_value_policy::copy,
                R"(
                 The shape of the output referred to by this output handle.
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/graph/node_output.hpp:83]: In my opinion, there should be a note about creating a copy of the Shape object. As for example `get_tensor` or `get_rt_info` are returning references -- so retrun-by-copy without the explicit note could create a misunderstanding in general.

CC @akuporos what do you think?

File: src/bindings/python/src/pyopenvino/graph/node_output.hpp [hunk 2]
```diff
@@ -88,6 +89,7 @@ void regclass_graph_Output(py::module m, std::string typestring)
                )");
     output.def("get_partial_shape",
                &ov::Output<VT>::get_partial_shape,
+               py::return_value_policy::copy,
                R"(
                 The partial shape of the output referred to by this output handle.
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/graph/node_output.hpp:94]: Same comment here.

File: src/bindings/python/src/pyopenvino/graph/node_output.hpp [hunk 3]
```diff
@@ -137,8 +139,8 @@ void regclass_graph_Output(py::module m, std::string typestring)
     output.def_property_readonly("any_name", &ov::Output<VT>::get_any_name);
     output.def_property_readonly("names", &ov::Output<VT>::get_names);
     output.def_property_readonly("element_type", &ov::Output<VT>::get_element_type);
-    output.def_property_readonly("shape", &ov::Output<VT>::get_shape);
-    output.def_property_readonly("partial_shape", &ov::Output<VT>::get_partial_shape);
+    output.def_property_readonly("shape", &ov::Output<VT>::get_shape, py::return_value_policy::copy);
+    output.def_property_readonly("partial_shape", &ov::Output<VT>::get_partial_shape, py::return_value_policy::copy);
```
Comment (jiwaszki) [inline src/bindings/python/src/pyopenvino/graph/node_output.hpp:143]: Should there be a note about copy?

File: src/bindings/python/tests/test_inference_engine/test_function.py [hunk 4]
```diff
@@ -349,3 +349,15 @@ def test_reshape_with_names():
...
         model.reshape({input.any_name: new_shape})
         assert input.partial_shape == new_shape
+
+
+def test_reshape(device):
+    shape = Shape([1, 10])
+    param = ops.parameter(shape, dtype=np.float32)
+    model = Model(ops.relu(param), [param])
+    ref_shape = model.input().partial_shape
+    ref_shape[0] = Dimension(3)
```
Comment (jiwaszki) [inline src/bindings/python/tests/test_inference_engine/test_function.py:359]:
```suggestion
    ref_shape[0] = 3
```
Is it possible?

Comment (AlexeyLebedev1) [reply src/bindings/python/tests/test_inference_engine/test_function.py:359]: No, but we can make it possible in python, adding one more overload of `__setitem__`

Comment (jiwaszki) [reply src/bindings/python/tests/test_inference_engine/test_function.py:359]: Please add it. I see it for `Shape` but there is no such overload for `PartialShape`...

Comment (AlexeyLebedev1) [reply src/bindings/python/tests/test_inference_engine/test_function.py]: Done.

File: src/bindings/python/src/pyopenvino/graph/node_output.hpp (hunk 1 above)
Comment (AlexeyLebedev1) [reply src/bindings/python/src/pyopenvino/graph/node_output.hpp:83]: I added a note about copy below. But anyway, c++ api also doesn't allow to change port's shape directly and i believe copying in this case is not heavy and for python user it shouldn't have a big difference here.

File: src/bindings/python/src/pyopenvino/graph/node_output.hpp (hunk 3 above)
Comment (AlexeyLebedev1) [reply src/bindings/python/src/pyopenvino/graph/node_output.hpp:143]: We can add docstring for all properties in a separate pr.

File: src/bindings/python/src/pyopenvino/graph/node_output.hpp (hunk 1 above)
Comment (jiwaszki) [reply src/bindings/python/src/pyopenvino/graph/node_output.hpp:83]: Sure, copy is fine here. I was only questioning if the note is necessary here, as for the user a possibility to change something could mean that the object reflects an actual state of internal values.

**Review by jiwaszki — APPROVED** [review]

**Pull request merged**